src/main/java/uk/gov/hmcts/reform/sandl/snlevents/actions/hearing/AssignSessionsToHearingAction.java

```diff
@@ -0,0 +1,161 @@
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.Hearing;
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.HearingPart;
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.Session;
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.UserTransactionData;
+import uk.gov.hmcts.reform.sandl.snlevents.model.request.HearingSessionRelationship;
+import uk.gov.hmcts.reform.sandl.snlevents.model.request.SessionAssignmentData;
+import uk.gov.hmcts.reform.sandl.snlevents.repository.db.HearingRepository;
+import uk.gov.hmcts.reform.sandl.snlevents.repository.db.SessionRepository;
+import uk.gov.hmcts.reform.sandl.snlevents.service.RulesService;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import javax.persistence.EntityManager;
+
+public class AssignSessionsToHearingAction extends Action implements RulesProcessable {
+
+    private final EntityManager entityManager;
+    protected HearingSessionRelationship relationship;
+    protected UUID hearingId;
+    protected Hearing hearing;
+    protected List<Session> targetSessions;
+    protected List<UUID> targetSessionsIds;
+    protected List<String> previousHearingParts;
+    protected String previousHearing;
+
+    protected HearingRepository hearingRepository;
+    protected SessionRepository sessionRepository;
+
+    public AssignSessionsToHearingAction(UUID hearingId,
+                                         HearingSessionRelationship relationship,
+                                         HearingRepository hearingRepository,
+                                         SessionRepository sessionRepository,
+                                         EntityManager entityManager,
+                                         ObjectMapper objectMapper) {
+        this.relationship = relationship;
+        this.hearingId = hearingId;
+        this.hearingRepository = hearingRepository;
+        this.sessionRepository = sessionRepository;
+        this.objectMapper = objectMapper;
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public void getAndValidateEntities() {
+        hearing = hearingRepository.findOne(hearingId);
+
+        targetSessionsIds = relationship.getSessionsData().stream()
+            .map(SessionAssignmentData::getSessionId)
+            .collect(Collectors.toList());
+        targetSessions = sessionRepository.findSessionByIdIn(targetSessionsIds);
+
+        if (targetSessions == null) {
+            throw new RuntimeException("Target sessions cannot be null!");
+        } else if (hearing == null) {
+            throw new RuntimeException("Hearing part cannot be null!");
+        }
+    }
+
+    @Override
+    public UUID[] getAssociatedEntitiesIds() {
+        final List<UUID> entitiesIds = new ArrayList<>();
+        entitiesIds.add(hearing.getId());
+        for (HearingPart hp : hearing.getHearingParts()) {
```

Comment:
why streams? ;)

Tried to be consistent with the rest of the file. For example, the act method had streams with forEach but then it would require internal (in foreach) try/catch of JsonProcessingException and another one for objectMappers.